src/core/thread/router_table.cpp

```diff
@@ -502,23 +502,25 @@ uint8_t RouterTable::GetLinkCost(Router &aRouter)
     return rval;
 }
 
-void RouterTable::ProcessTlv(const Mle::RouteTlv &aTlv)
+void RouterTable::ProcessRouterIdSet(uint8_t aRouterIdSequence, const Mle::RouterIdSet &aRouterIdSet)
 {
     bool allocationChanged = false;
 
-    mRouterIdSequence            = aTlv.GetRouterIdSequence();
+    mRouterIdSequence            = aRouterIdSequence;
     mRouterIdSequenceLastUpdated = TimerMilli::GetNow();
 
+    VerifyOrExit(mAllocatedRouterIds != aRouterIdSet);
+
     for (uint8_t routerId = 0; routerId <= Mle::kMaxRouterId; routerId++)
     {
-        if (aTlv.IsRouterIdSet(routerId) == IsAllocated(routerId))
+        if (aRouterIdSet.Contains(routerId) == IsAllocated(routerId))
         {
             continue;
         }
 
         allocationChanged = true;
 
-        if (aTlv.IsRouterIdSet(routerId))
+        if (aRouterIdSet.Contains(routerId))
         {
             mAllocatedRouterIds.Add(routerId);
```

Comment:
How about if remove the `mAllocatedRouterIds.Add(routerId)` and the `mAllocatedRouterIds.Remove(routerId);` in the `else`?  And instead after the loop, do a `mAllocatedRouterIds = aRouterIdSet` to copy all the bits at the same time.

This should hep make the the loop simple: We go through all the router id values and check the ones that are removed (are contained in the old `mAllocatedRouterIds` but not in the new `aRouterIdSet`) and do the `SetNextHop` and `RemoveRouterLink` for those...

Reply:
good point!